pkg/plugins/gateway/algorithms/least_request.go

```diff
@@ -112,8 +76,41 @@ func (r leastRequestRouter) Route(ctx *types.RoutingContext, pods types.PodList)
 
 func (r *leastRequestRouter) SubscribedMetrics() []string {
 	return []string{
-		metrics.NumRequestsRunning,
-		metrics.NumRequestsWaiting,
-		metrics.NumRequestsSwapped,
+		metrics.RealtimeNumRequestsRunning,
+	}
+}
+
+func selectTargetPodWithLeastRequestCount(cache cache.Cache, readyPods []*v1.Pod) *v1.Pod {
+	var targetPod *v1.Pod
+	targetPods := []string{}
+
+	minCount := math.MaxInt32
+	podRequestCount := getRequestCounts(cache, readyPods)
+	for _, totalReq := range podRequestCount {
+		if totalReq <= minCount {
+			minCount = totalReq
+		}
 	}
+	for podname, totalReq := range podRequestCount {
+		if totalReq == minCount {
+			targetPods = append(targetPods, podname)
+		}
+	}
+	if len(targetPods) > 0 {
+		targetPod, _ = utils.FilterPodByName(targetPods[rand.Intn(len(targetPods))], readyPods)
+	}
+	return targetPod
+}
+
+func getRequestCounts(cache cache.Cache, readyPods []*v1.Pod) map[string]int {
+	podRequestCount := map[string]int{}
+	for _, pod := range readyPods {
+		runningReq, err := cache.GetMetricValueByPod(pod.Name, metrics.RealtimeNumRequestsRunning)
```

Comment:
do you observe any issue on waiting and swapped?  why are they removed from the list?

Reply:
Tracking running request count within gateway was introduced in PR: #918. Changed least_request to use running_requests tracked by gateway and remove dependency of fetching metrics (running, swapped and waiting) from vllm. 

Reason to not use vllm metrics is that it is prone to burst requests.

Reply:
1) I validated the correctness for tracking running requests within gateway cache vs source of truth (inference engine).

2) For stateful use cases such as tracking running requests or prefix hash blocks, only one gateway leader instances should be used for accuracy. For running multiple gateway instances in all active pattern, requires backend storage implementation to sync state across gateway instances: https://github.com/vllm-project/aibrix/issues/761

Reply:
I recommend to leave a note here.  This needs to be refactored at the time we moved to HA case

Reply:
Done.